mypy/checkmember.py

```diff
@@ -187,7 +187,53 @@ def analyze_member_var_access(name: str, itype: Instance, info: TypeInfo,
     if isinstance(vv, Decorator):
         # The associated Var node of a decorator contains the type.
         v = vv.var
+
     if isinstance(v, Var):
+        vtype = v.type
+        if isinstance(vtype, Instance):
+            dunder_get = vtype.type.get_method('__get__')
+            dunder_set = vtype.type.get_method('__set__')
+            dunder_delete = vtype.type.get_method('__delete__')
+            if dunder_get or dunder_set or dunder_delete:
+                if is_lvalue:
+                    if dunder_set or dunder_delete:
+                        if dunder_set:
+                            typ = map_instance_to_supertype(vtype, dunder_set.info)
+                            dunder_set_type = expand_type_by_instance(
+                                method_type_with_fallback(
+                                    dunder_set, builtin_type('builtins.function')
+                                ),
+                                typ
+                            )
+                            if isinstance(dunder_set_type, CallableType):
+                                return dunder_set_type.arg_types[1]
+                        else:
+                            msg.fail("Can't set a data descriptor with no __set__ method", node)
+                    else:
+                        # TODO: Capture setting a non-data descriptor to infer types
+                        # later
+                        return AnyType()
+                else:
+                    # this is a descriptor, use the output type of the __get__ method
+                    if dunder_get:
+                        typ = map_instance_to_supertype(vtype, dunder_get.info)
+                        dunder_get_type = expand_type_by_instance(
+                            method_type_with_fallback(
+                                dunder_get, builtin_type('builtins.function')
+                            ),
+                            typ
+                        )
+                        if isinstance(dunder_get_type, CallableType):
+                            return dunder_get_type.ret_type
+                        else:
+                            return AnyType()
+                    else:
+                        # TODO: Per the docs: "If [a descriptor] does not define __get__(),
+                        # then accessing the attribute will return the descriptor object
+                        # itself unless there is a value in the object’s instance dictionary".
```

Comment:
Ah, that's true. I'll take a stab at making it work (since it may just require falling through to the rest of the typing code).